scope.go

```diff
@@ -162,118 +154,149 @@ func (s *standardScope) reportLoop(interval time.Duration) {
-			val = &gauge{}
+			val = newGauge()
 			s.gauges[name] = val
 		}
 		s.gm.Unlock()
 	}
 	return val
 }
 
-// Timer returns the timer identified by the scope and the provided name, creating it if it does not already exist
-func (s *standardScope) Timer(name string) Timer {
+func (s *scope) Timer(name string) Timer {
 	s.tm.RLock()
 	val, ok := s.timers[name]
 	s.tm.RUnlock()
 	if !ok {
 		s.tm.Lock()
 		val, ok = s.timers[name]
 		if !ok {
-			val = &timer{
-				name:     s.fullyQualifiedName(name),
-				tags:     s.tags,
-				reporter: s.reporter,
-			}
+			val = newTimer(s.fullyQualifiedName(name), s.tags, s.reporter)
 			s.timers[name] = val
 		}
 		s.tm.Unlock()
 	}
 	return val
 }
 
-// mergeRightTags merges 2 sets of tags with the tags from tagsRight overriding values from tagsLeft
-func mergeRightTags(tagsLeft, tagsRight map[string]string) map[string]string {
-	if tagsLeft == nil && tagsRight == nil {
-		return nil
-	}
-
-	result := make(map[string]string, len(tagsLeft)+len(tagsRight))
-	for k, v := range tagsLeft {
-		result[k] = v
-	}
-	for k, v := range tagsRight {
-		result[k] = v
-	}
-	return result
-}
-
-func (s *standardScope) Tagged(tags map[string]string) Scope {
-	subscope := &standardScope{
+func (s *scope) Tagged(tags map[string]string) Scope {
+	subscope := &scope{
 		prefix:   s.prefix,
 		tags:     mergeRightTags(s.tags, tags),
 		reporter: s.reporter,
 		registry: s.registry,
 
-		counters: make(map[string]Counter),
-		gauges:   make(map[string]Gauge),
-		timers:   make(map[string]Timer),
+		counters: make(map[string]*counter),
+		gauges:   make(map[string]*gauge),
+		timers:   make(map[string]*timer),
 	}
 
 	subscope.registry.add(subscope)
 	return subscope
 }
 
-func (s *standardScope) SubScope(prefix string) Scope {
-	subscope := &standardScope{
+func (s *scope) SubScope(prefix string) Scope {
+	subscope := &scope{
 		prefix:   s.fullyQualifiedName(prefix),
 		tags:     s.tags,
 		reporter: s.reporter,
 		registry: s.registry,
 
-		counters: make(map[string]Counter),
-		gauges:   make(map[string]Gauge),
-		timers:   make(map[string]Timer),
+		counters: make(map[string]*counter),
+		gauges:   make(map[string]*gauge),
+		timers:   make(map[string]*timer),
 	}
 
 	subscope.registry.add(subscope)
 	return subscope
 }
 
-func (s *standardScope) Reporter() StatsReporter {
-	return s.reporter
+func (s *scope) Capabilities() Capabilities {
+	if s.reporter == nil {
+		return capabilitiesNone
+	}
+	return s.reporter.Capabilities()
+}
+
+func (s *scope) Snapshot() Snapshot {
+	snap := newSnapshot()
+
+	s.registry.sm.RLock()
+	for _, ss := range s.registry.subscopes {
+		// NB(r): tags are immutable, no lock required to read.
+		tags := make(map[string]string, len(s.tags))
+		for k, v := range ss.tags {
+			tags[k] = v
+		}
+
+		ss.cm.RLock()
+		for key, c := range ss.counters {
+			name := ss.fullyQualifiedName(key)
+			snap.counters[name] = &counterSnapshot{
+				name:  name,
+				tags:  tags,
+				value: c.snapshot(),
```

Comment:
snapshot doesn't update the prev value, so for counters, snapshot is always going to be ever increasing. Is this the expected behavior or do we want snapshot to only have delta between snapshot calls?